Fix IO throttling during copying whole directories

azat opened the pull request:
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix IO throttling during copying whole directories (MOVE PARTITION TO VOLUME, fetching/cloning parts, mutations, ...)

**P.S. help with tests is highly appreciated** (_for now I've tested only locally `MOVE PARTITION TO VOLUME`_)

Fixes: #52753 (cc @den-crane)
Fixes: #52183 (cc @coulsonm)

azat committed:
Fix IO throttling during copying whole directories

v2: pass local context during ALTER TABLE MOVE PARTITION TO VOLUME/DISK as well
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #53338 from azat/ClickHouse:throttling-fixes -->
This is an automated comment for commit d349988838ae09e9c27966e8671a5a72892e2405 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53338/d349988838ae09e9c27966e8671a5a72892e2405/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

coulsonm commented:
![image](https://github.com/ClickHouse/ClickHouse/assets/127475879/fc38577c-66b3-490a-ac82-c375e82b3b21)

FileNotFoundError: [Errno 2] No such file or directory: '/home/ubuntu/actions-runner/_work/_temp/tests_bugfix_check/integration/integration_commit_status.tsv'

coulsonm commented:
@azat

azat commented:
This is likely internal CI issue. And this PR does not have a test yet.

coulsonm commented:
May I ask when will it be repaired?

azat commented:
This should not be required for merging this PR, it just needs some attention. @serxa maybe you can take a look?

serxa commented on src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:
```diff
@@ -769,8 +769,11 @@ void DiskObjectStorageTransaction::createFile(const std::string & path)
+void DiskObjectStorageTransaction::copyFile(const std::string & from_file_path, const std::string & to_file_path, const WriteSettings & settings)
 {
+    /// NOTE: For native copy we can ignore throttling, so no need to use WriteSettings
+    (void)settings;
```
```suggestion
    UNUSED(settings);
```

serxa reviewed (approved):
Looks good. As for tests, at least a simple check for the duration of explicit ALTER TABLE MOVE PARTITION would be good enough.

azat committed:
Add a test for IO throttling for ALTER TABLE MOVE PARTITION TO VOLUME

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
> As for tests, at least a simple check for the duration of explicit ALTER TABLE MOVE PARTITION would be good enough.

Yeah, I was just lazy and thought maybe someone will contribute the test in the mean time, but this did not happen, so I've added the test.

azat committed:
Add no-replicated-database for 02870_move_partition_to_volume_io_throttling

ALTER TABLE MOVE PARTITION TO DISK/VOLUME should not be replicated since
replicas will not contains this part:

    azat:~/ch/tmp/53338$ for i in clickhouse-server*.log.zst; do echo -n "$i: " && zstd -cdq $i | grep -m1 -e Executed.*ALTER.*test_cnmf4xnb.test_move_partition_throttling -e Exception.*ALTER.*test_cnmf4xnb.test_move_partition_throttling; done
    clickhouse-server1.log.zst: 2023.08.29 16:46:53.960065 [ 1843 ] {19b13d67-54c0-496d-96b9-b1ec09df3618} <Error> executeQuery: Code: 232. DB::Exception: Nothing to move (check that the partition exists). (NO_SUCH_DATA_PART) (version 23.8.1.2862) (from 0.0.0.0:0) (comment: 02870_move_partition_to_volume_io_throttling.sql) (in query: /* ddl_entry=query-0000000005 */ ALTER TABLE test_cnmf4xnb.test_move_partition_throttling MOVE PARTITION tuple() TO VOLUME 'remote' SETTINGS max_remote_write_network_bandwidth = 1600000), Stack trace (when copying this message, always include the lines below):
    clickhouse-server2.log.zst: 2023.08.29 16:46:53.959560 [ 1842 ] {3cd2b5e8-24b9-4cfd-aa47-854e634936f2} <Error> executeQuery: Code: 232. DB::Exception: Nothing to move (check that the partition exists). (NO_SUCH_DATA_PART) (version 23.8.1.2862) (from 0.0.0.0:0) (comment: 02870_move_partition_to_volume_io_throttling.sql) (in query: /* ddl_entry=query-0000000005 */ ALTER TABLE test_cnmf4xnb.test_move_partition_throttling MOVE PARTITION tuple() TO VOLUME 'remote' SETTINGS max_remote_write_network_bandwidth = 1600000), Stack trace (when copying this message, always include the lines below):
    clickhouse-server.log.zst: 2023.08.29 16:46:53.950730 [ 721 ] {df6b20ee-1903-404a-9398-7fead209ccd7} <Debug> DDLWorker(test_cnmf4xnb): Executed query: /* ddl_entry=query-0000000005 */ ALTER TABLE test_cnmf4xnb.test_move_partition_throttling MOVE PARTITION tuple() TO VOLUME 'remote' SETTINGS max_remote_write_network_bandwidth = 1600000

CI: https://s3.amazonaws.com/clickhouse-test-reports/53338/ed401cba9b8254b4a29a5ed9e5ad838c26ffaac1/stateless_tests__release__databasereplicated__[4_4].html
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>
